### DISCUSSION TITLE
Notarizing the captured photos and videos in Mapeo

### DISCUSSION THREAD

**Discussion by arky** [digidem/mapeo-mobile#769] [Ideas]
Is there way to notarize photos and videos in Mapeo. The notarization process ensures that the GPS location and other metadata could be verified at a later date. 

CameraV approach is described here:
https://guardianproject.github.io/informacam-guide/en/InformacamGuide.html

**Comment by gmaclennan** [comment]
Hi @arky yes we've been doing some thinking about this. We already have some of this, but no UI to access it:

Internally when you create an observation it is written to a [`hypercore`](https://github.com/mafintosh/hypercore). This is an append-only log and every entry is cryptographically signed, so it cannot be changed later and it can be proved to come from a specific device. The observations contain references to the photos as well as the GPS coordinates, so they are a form of "proof". However there are currently a couple of weaknesses:

1. A user could edit / replace a photo with a another with the same filename - the observation only references the photo by filename so there is no guarantee the photo has not been changed after, although it is not possible to do this in the app - it would require tampering with the data after export or from desktop.

2. There is no guarantee that the GPS coordinates are correct. If the GPS is switched off then Mapeo allows the user to manually enter coordinates. This is recorded in the observation metadata, but a bad actor could create an observation on Desktop with whatever data they want.

(1) can be solved by storing a hash of the photo in the observation record. The observation record cannot be tampered with, so you could not change the hash, and you could detect if the image was changed (hash of the image would not match hash recorded on the observation).

(2) needs a solution like CameraV - recording sensor data which would be hard / impossible to fake. One way of doing this would be to record raw GPS data in observation metadata - I've tested out getting a feed of raw NMEA sentences, which include details of which satellites are in view and what angle they are at. We could record these to the observation whilst the user is taking a photo. These would be extremely hard to fake - but also very hard to verify. I'm not sure how you can look up and verify where all the GPS satellites were at a particular moment in time. But I think that's the case with the CameraV's methods as well - the harder they are to fake, the harder they are to verify.

I think recording NMEA sentences would act as reasonable proof, combined with hashing and the existing signatures we already have on observations, we can show that an observation was taken at a particular moment, the photo was taken at the same moment, and it hasn't been tampered with since, and show evidence with the NMEA sentences that the user was at a particular location - it would be the equivalent (better) of taking a picture of a GPS screen.
